ocaml/xapi/xapi_session.ml

```diff
@@ -27,6 +27,235 @@ open Client
+
+  let client_of_info ~__context ~originator ~uname =
+    let user_agent = Context.get_user_agent __context in
+    let ip = Context.get_client_ip __context in
+    (* check to make sure we have at least _some_ information *)
+    if
+      [user_agent; originator; uname; ip]
+      |> List.for_all (function None | Some "" -> true | _ -> false)
+    then
+      None
+    else
+      Some {originator; uname; user_agent; ip}
+
+  let string_of_client x =
+    [
+      ("username", x.uname)
+    ; ("originator", x.originator)
+    ; ("useragent", x.user_agent)
+    ; ("ip", x.ip)
+    ]
+    |> List.filter_map (fun (label, value) ->
+           match value with
+           | None | Some "" ->
+               None
+           | Some value ->
+               Some (Printf.sprintf "<%s>%s</%s>" label value label))
+    |> String.concat "\n"
+
+  type client_failed_attempts = {
+      client: client
+    ; num_failed_attempts: int
+    ; last_failed_attempt: Date.iso8601
+  }
+
+  let up_to_3 xs x =
+    let rec merge = function
+      | [] ->
+          [x]
+      | y :: ys ->
+          if x.num_failed_attempts >= y.num_failed_attempts then
+            x :: y :: ys
+          else
+            y :: merge ys
+    in
+    let rec keep ctr xs =
+      match (ctr, xs) with
+      | ctr, x :: xs when ctr > 0 ->
+          x :: keep (ctr - 1) xs
+      | _ ->
+          []
+    in
+    merge xs |> keep 3
+
+  let string_of_client_failed_attempts x =
+    Printf.sprintf {|
+<known>
+%s
+<number>%i</number>
+<date>%s</date>
+</known>|}
+      (string_of_client x.client)
+      x.num_failed_attempts
+      (Date.to_string x.last_failed_attempt)
+
+  type stats = {
+      total_num_failed_attempts: int
+    ; top_3_worst_clients: client_failed_attempts list
+          (* not necessarily 3, but <=3 *)
+    ; unknown_client_failed_attempts: int
+  }
+
+  let string_of_stats
+      {
+        total_num_failed_attempts
+      ; top_3_worst_clients
+      ; unknown_client_failed_attempts
+      } =
+    let unknown =
+      if unknown_client_failed_attempts = 0 then
+        ""
+      else
+        Printf.sprintf {|
+<unknown>%i</unknown>|} unknown_client_failed_attempts
+    in
+    let known_with_total =
+      if List.length top_3_worst_clients = 0 then
+        ""
+      else
+        Printf.sprintf {|
+<total>%i</total>%s|} total_num_failed_attempts
+          (top_3_worst_clients
+          |> List.map string_of_client_failed_attempts
+          |> String.concat ""
+          )
+    in
+    Printf.sprintf {|<body>%s%s
+</body>|} known_with_total unknown
+
+  module Stats : sig
+    val get : unit -> stats option
+
+    (* returns the number of failures from this client since last call to [ get ] *)
+    val record_client : client -> now:Date.iso8601 -> int
+
+    (* returns number of failures from unknown clients since last call to [ get ] *)
+    val record_unknown : unit -> int
+  end = struct
+    let m = Mutex.create ()
+
+    let unknown_ctr = ref 0
```

Comment:
The counter and the Hashtbl could have be put into a record to stress the connection. However, this is a small module and fine as is.